KAFKA-15913: Migrate async consumer tests to mocks

lucasbru opened the pull request:
Use mocks to test the AsyncKafkaConsumer

 - Eliminate the use of ConsumerTestBuilder
 - Mock all resources that were previously retrieved via leaking
   the background thread with mockito spys
 - Always use the default constructor of AsyncKafkaConsumer as
   much as possible, inject mocks via factories.
 - Timeouts are mocked directly by timeout exceptions instead
   of waiting for futures to time out.

I did not port the autocommit mocking code, because it was mostly
testing the integration of foreground and background thread (or
making the spy's work which broke during the autocommit on close)
and is currently being reimplemented anyway.

New test runs 10x faster.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

lucasbru committed:
migrate epoch update to mocks

lucasbru committed:
migrate fenced exception

lucasbru committed:
migrate committed tests

lucasbru committed:
migrate subscribed tests

lucasbru committed:
migrate offset tests

lucasbru committed:
migrate wakeup tests

lucasbru committed:
migrate assing tests

cadonna committed:
Add test testCommitAsync_NullCallback as testCommitAsyncWithNullCallback

lucasbru committed:
use most of the normal construction code for testing and migrate config tests

lucasbru committed:
fix rebase errors, port remote assignor and group metadata tests

lucasbru committed:
port long poll test

lucasbru committed:
migrate commit callback tests

lucasbru committed:
migrate wakeup committed

lucasbru committed:
reorder

lucasbru committed:
updates

lucasbru committed:
checkstyle

cadonna commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java:
```diff
@@ -16,8 +16,43 @@
  */
 package org.apache.kafka.clients.consumer.internals;
 
+import static java.util.Arrays.asList;
+import static java.util.Collections.singleton;
+import static java.util.Collections.singletonList;
+import static org.apache.kafka.clients.consumer.internals.ConsumerUtils.THROW_ON_FETCH_STABLE_OFFSET_UNSUPPORTED;
+import static org.apache.kafka.common.utils.Utils.mkEntry;
+import static org.apache.kafka.common.utils.Utils.mkMap;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.atLeast;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
```
Static imports should be the last block of imports.

lucasbru replied:
Done, but source? Where do we define these thingds

cadonna replied:
Common knowledge 🙂 
Jokes aside, I do not know where and if we specify it somewhere in written form.

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -482,6 +497,47 @@ private void process(final GroupMetadataUpdateEvent event) {
     }
 
+    // auxiliary interface for testing
+    interface ApplicationEventHandlerSupplier {
```
should also be called factory

lucasbru replied:
Done

cadonna commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java:
```diff
@@ -42,192 +77,142 @@
+        return setup(config);
+    }
+
+    private AsyncKafkaConsumer<String, String> setup(ConsumerConfig config) {
```
nit: 
Setup is quite generic. What about:
```suggestion
    private AsyncKafkaConsumer<String, String> newConsumer(ConsumerConfig config) {
```

lucasbru replied:
Done

cadonna commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java:
```diff
@@ -42,192 +77,142 @@
     @MethodSource("commitExceptionSupplier")
-    public void testCommitAsync_UserSuppliedCallback(Exception exception) {
-        CompletableFuture<Void> future = new CompletableFuture<>();
+    public void testCommitAsyncUserSuppliedCallback(Exception exception) {
```
I would separate the exceptional from the non-exceptional case. It would make the test more readable.

lucasbru replied:
Done

cadonna commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java:
```diff
@@ -42,192 +77,142 @@
-        doReturn(future).when(consumer).commit(offsets, false);
         MockCommitCallback callback = new MockCommitCallback();
         assertDoesNotThrow(() -> consumer.commitAsync(offsets, callback));
+        consumer.maybeInvokeCommitCallbacks();
```
As far as, I understand the agreement is that the commit callback is called in `poll()`, `close()`, `commitSync()`, and `commitAsync()`. Why make `maybeInvokeCommitCallbacks()` visible instead of calling one of the mentioned methods?

lucasbru replied:
Done

cadonna commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java:
```diff
@@ -1113,45 +1053,41 @@ private HashMap<TopicPartition, Long> mockTimestampToSearch() {
-                (short) 1,
-                Errors.NONE).responseBody());
-        }
+    private void mockCommitApplicationEventException(Exception ex) {
```
I think calling this method (and its siblings) something like `completeCommitApplicationEventExceptionally()` (`completeCommitApplicationEventSuccessfully()` for the good case) would make the code better readable.

cadonna commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java:
```diff
@@ -241,65 +226,64 @@ private static Stream<Exception> commitExceptionSupplier() {
 
     @Test
     public void testFencedInstanceException() {
```
I would change the name to "testCommitAsyncWithFencedException".

lucasbru replied:
Done

cadonna commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java:
```diff
@@ -241,65 +226,64 @@ private static Stream<Exception> commitExceptionSupplier() {
+
         assertDoesNotThrow(() -> consumer.commitAsync());
-        future.completeExceptionally(Errors.FENCED_INSTANCE_ID.exception());
     }
```
Isn't here something missing? The next call to `consumer.commitAsync()` should throw, right? Shouldn't we verify that?

lucasbru replied:
Good point. There were a couple of poor tests and I didn't feel like fixing it all, but this one is a good one to fix. Done

cadonna commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java:
```diff
@@ -241,65 +226,64 @@ private static Stream<Exception> commitExceptionSupplier() {
-            verify(applicationEventHandler).add(ArgumentMatchers.isA(FetchCommittedOffsetsApplicationEvent.class));
-        }
+        assertDoesNotThrow(() -> consumer.committed(topicPartitionOffsets.keySet(), Duration.ofMillis(1000)));
+        verify(applicationEventHandler).addAndGet(ArgumentMatchers.isA(FetchCommittedOffsetsApplicationEvent.class), any());
```
I am missing the verification of the return value of `committed()`.

lucasbru replied:
Also didn't have it before but I fixed it.

cadonna commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -482,6 +497,47 @@ private void process(final GroupMetadataUpdateEvent event) {
     }
 
+    // auxiliary interface for testing
+    interface ApplicationEventHandlerSupplier {
```
It is nor clear to me why we need a supplier here. Could you elaborate?

lucasbru replied:
I want to inject a custom implementation of the applicationEventHandler. It needs to work for both prod and testing so that we can actually test the code we are running. The instantiation for prod depends on arguments that are created inside the constructor of AsyncKafkaConsumer. If I do not define this supplier, I cannot inject the constructor of the "real" ApplicationEventHandler for prod.

cadonna commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java:
```diff
@@ -505,57 +484,10 @@ public void onComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Exception
-     * returning normally, timing out, throwing an error, etc. By mocking the construction of the event object that
-     * is created, we can affect that behavior.
-     */
-    private static MockedConstruction<FetchCommittedOffsetsApplicationEvent> offsetFetchEventMocker(CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> future) {
-        // This "answer" is where we pass the future to be invoked by the ConsumerUtils.getResult() method
-        Answer<Map<TopicPartition, OffsetAndMetadata>> getInvocationAnswer = invocation -> {
-            // This argument captures the actual argument value that was passed to the event's get() method, so we
-            // just "forward" that value to our mocked call
-            Timer timer = invocation.getArgument(0);
-            return ConsumerUtils.getResult(future, timer);
-        };
-
-        MockedConstruction.MockInitializer<FetchCommittedOffsetsApplicationEvent> mockInitializer = (mock, ctx) -> {
-            // When the event's get() method is invoked, we call the "answer" method just above
-            when(mock.get(any())).thenAnswer(getInvocationAnswer);
-
-            // When the event's type() method is invoked, we have to return the type as it will be null in the mock
-            when(mock.type()).thenReturn(ApplicationEvent.Type.FETCH_COMMITTED_OFFSETS);
-
-            // This is needed for the WakeupTrigger code that keeps track of the active task
-            when(mock.future()).thenReturn(future);
-        };
-
-        return mockConstruction(FetchCommittedOffsetsApplicationEvent.class, mockInitializer);
-    }
-
-    private static MockedConstruction<CommitApplicationEvent> commitEventMocker(CompletableFuture<Void> future) {
-        Answer<Void> getInvocationAnswer = invocation -> {
-            Timer timer = invocation.getArgument(0);
-            return ConsumerUtils.getResult(future, timer);
-        };
-
-        MockedConstruction.MockInitializer<CommitApplicationEvent> mockInitializer = (mock, ctx) -> {
-            when(mock.get(any())).thenAnswer(getInvocationAnswer);
-            when(mock.type()).thenReturn(ApplicationEvent.Type.COMMIT);
-            when(mock.future()).thenReturn(future);
-        };
-
-        return mockConstruction(CommitApplicationEvent.class, mockInitializer);
-    }
```
🥳

cadonna commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -1535,6 +1591,11 @@ int callbacks() {
         return invoker.callbackQueue.size();
     }
 
+    // Visible for testing
+    SubscriptionState subscriptions() {
+        return subscriptions;
+    }
```
Is there a way to avoid this?

lucasbru replied:
- I can add callback to the constructor that registers the SubscriptionState with the calling context (the test)
- I can add another factory to inject our own instance of SubscriptionState

cadonna reviewed (commented):
Thanks for the PR, @lucasbru !

Here my feedback.

lucasbru committed:
PR comments

lucasbru committed:
Merge remote-tracking branch 'upstream/trunk' into ctr_migrate_test

cadonna reviewed (approved):
LGTM!

We can work on the open test-specific questions in a follow-up PR. Now it is important to get a first version of the spy-free unit tests merged.

Thanks, @lucasbru !

lucasbru committed:
remove unused code.
make fenced exception test more unflaky.
assert( ... instanceof ... ) -> assertIsInstanceOf(...)

lucasbru committed:
Merge remote-tracking branch 'upstream/trunk' into ctr_migrate_test

lucasbru committed:
optimize imports

lucasbru committed:
checkstyle

lucasbru committed:
Merge remote-tracking branch 'upstream/trunk' into ctr_migrate_test

lucasbru committed:
merge fixes
